pkg/services/ngalert/state/cache.go

```diff
@@ -253,6 +249,25 @@ func (c *cache) getStatesForRuleUID(orgID int64, alertRuleUID string) []*State {
 	return result
 }
 
+func (c *cache) forRuleState(orgID int64, alertRuleUID string, fn func(*State)) {
+	c.mtxStates.RLock()
+	defer c.mtxStates.RUnlock()
+
+	orgRules, ok := c.states[orgID]
+	if !ok {
+		return
+	}
+
+	rs, ok := orgRules[alertRuleUID]
+	if !ok {
+		return
+	}
+
+	for _, state := range rs.states {
+		fn(state)
```

Comment:
Perhaps it makes sense to encapsulate deletion of stale state here? This would be more atomic operation.